tests/authenticator/RESTfulAPI_TokenAuthenticator_Test.php

```diff
@@ -209,4 +214,84 @@ public function testAuthenticate()
       "TokenAuth authentication failure should return a RESTfulAPI_Error"
     );
   }
-}
\ No newline at end of file
+
+  public function testCanLogin(){
+    $auth = $this->getAuthenticator();
+
+    // check login when canLogIn check fails
+    $request = new SS_HTTPRequest(
+      'GET',
+      'api/auth/login',
+      array(
+        'email' => 'balrog@moria.mine',
+        'pwd'   => 'flame.sword'
+      )
+    );
+
+    $result = $auth->login($request);
+
+    $this->assertEquals(
+      $result['code'],
+      RESTfulAPI_TokenAuthenticator::AUTH_CODE_LOGIN_FAIL,
+      "Logging in a member whose `canLogIn` check fails should not be allowed."
+    );
+
+    // check authenticate when canLogIn check fails
+
+    $request = new SS_HTTPRequest(
+      'GET',
+      'api/auth/login',
+      array(
+        'email' => 'test@test.com',
+        'pwd'   => 'test'
+      )
+    );
+
+    $result = $auth->login($request);
+
+    $this->assertEquals(
+      $result['code'],
+      RESTfulAPI_TokenAuthenticator::AUTH_CODE_LOGGED_IN,
+      "Logging in a member whose `canLogIn` check doesn't fail should be allowed."
+    );
+
+    $request = new SS_HTTPRequest(
+      'GET',
+      'api/ApiTest_Book/1'
+    );
+    $request->addHeader('X-Silverstripe-Apitoken', $result['token']);
+
+    // deny all users for the authenticate request
+    TestCanLoginExtension::$denyAll = true;
+
+    $result = $auth->authenticate($request);
+
+    TestCanLoginExtension::$denyAll = false;
+
+    $this->assertContainsOnlyInstancesOf(
+      'RESTfulAPI_Error',
+      array($result),
+      "TokenAuth authentication should fail when 'canLogIn' fails"
+    );
+  }
+}
+
+/**
+ * Extension to test "canLogin"
+ */
+class TestCanLoginExtension extends DataExtension
+{
+  public static $denyAll = false;
+
+  public function canLogIn(ValidationResult &$result)
+  {
+    if(self::$denyAll){
```

Comment:
Do we need all this login? We could just have one line like `$result->error('You shall not pass!');` which would have the same effect.. this would simplify the whole thing.